tony-core/src/main/java/com/linkedin/tony/util/Utils.java

```diff
@@ -374,13 +391,34 @@ public static void addEnvironmentForResource(LocalResource resource, FileSystem
       if (numInstances > 0) {
         // We rely on unique priority behavior to match allocation request to task in Hadoop 2.7
         containerRequests.put(jobName,
-                new TensorFlowContainerRequest(jobName, numInstances, memory, vCores, gpus, priority, nodeLabel));
+                new JobContainerRequest(jobName, numInstances, memory, vCores, gpus, priority,
+                    nodeLabel, dependsOn));
         priority++;
       }
     }
     return containerRequests;
   }
 
+  private static void ensureStagedTasksIntegrity(List<String> prepareStageTasks, List<String> trainingStageTasks,
+      Set<String> allJobTypes) {
+    if (prepareStageTasks.isEmpty() && !trainingStageTasks.isEmpty()) {
+      prepareStageTasks.addAll(CollectionUtils.subtract(allJobTypes, trainingStageTasks));
+      LOG.warn("Found no prepare stage tasks, auto-filling with: " + Arrays.toString(prepareStageTasks.toArray()));
+    } else if ((!prepareStageTasks.isEmpty() && trainingStageTasks.isEmpty())) {
+      trainingStageTasks.addAll(CollectionUtils.subtract(allJobTypes, prepareStageTasks));
+      LOG.warn("Found no training stage tasks, auto-filling with: " + Arrays.toString(trainingStageTasks.toArray()));
+    } else if (prepareStageTasks.isEmpty() && trainingStageTasks.isEmpty()) {
+      return;
+    }
+
+    if (prepareStageTasks.size() + trainingStageTasks.size() != allJobTypes.size()) {
```

Comment:
See the other comment, Do we need to enforce this invariant? IMO we should just enforce the user's configured 'training' tasks depend on user's configured 'prepare' tasks, everything else can just be scheduled as normal

Reply:
I think this makes sense, I'll remove the training-stage api. Just having prepare-stage is enough

Reply:
Discussed offline...it's best to enforce this ( prepare jobs + training jobs = all jobs). Otherwise if a job is  misconfigured (e.g. a job that was supposed to be in training phase, but was not configured as such) will be allocated first and will sit idle (wasting cluster resources).